Authentication & Authorisation based on jwt token

bhtibrewal opened the pull request:
<!-- Date Here in dd mmm yyyy-->
Date: 3-3-2024
<!--Developer Name Here-->
Developer Name: @bhtibrewal 

---

## Issue Ticket Number
https://github.com/Real-Dev-Squad/skill-tree-backend/issues/16

## Description
 - using the rds-v2-session cookie received in the API calls 
 - check if the cookie is valid using the RDS_PUBLIC_KEY 
 - get the rdsUserId and role from this cookie to authenticate & authorise the user 
 - add authentication & authorisation checks on routes in the security config 
 
 - update the existing integration tests to use the cookie
 - added tests for security checks

### Documentation Updated?
Documentation for the changes in this PR was already added in the README previously
- [ ] Yes
- [x] No

<!--Additional notes about documentation update if applicable-->

### Under Feature Flag

- [ ] Yes
- [x] No

<!--Indicate if changes are under a feature flag-->

### Database Changes

- [ ] Yes
- [x] No

<!--Notes on any database changes-->

### Breaking Changes

- [x] Yes
- [ ] No

<!--Notes on breaking changes or related issue tickets if applicable-->

### Development Tested?

- [x] Yes
- [ ] No

<!--Confirmation of local testing during development-->

## Screenshots
<details>
<summary>Valid Token </summary>

![image](https://github.com/Real-Dev-Squad/skill-tree-backend/assets/42600164/c3477074-fc4b-4e76-b4ac-f5ed96154668)

</details>

<details>
<summary>Invalid Token </summary>


![image](https://github.com/Real-Dev-Squad/skill-tree-backend/assets/42600164/68976510-95cd-46ae-b376-8fd248a82ff9)

<!-- Attach your screenshots here👇-->

</details>

<!--Attach or link to relevant screenshots or visual aids-->

## Test Coverage
<details>
<summary>Screenshot 1</summary>

![image](https://github.com/Real-Dev-Squad/skill-tree-backend/assets/42600164/b93aae70-c53b-4601-8bc8-1fcb1098d910)

</details>

<!--Attach Details on test coverage and outcomes-->

## Additional Notes

<!--Any additional notes, considerations or explanations for reviewers-->

debanjanc01 commented on skill-tree/pom.xml:
```diff
@@ -89,6 +90,31 @@
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-api</artifactId>
+            <version>0.11.2</version>
```
```suggestion
            <version>${io.jsonwebtoken.version}</version>
```

debanjanc01 commented on skill-tree/src/main/java/com/RDS/skilltree/Authentication/UserAuthenticationToken.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public Object getPrincipal(){
+        return UserModel.builder().rdsUserId(rdsUserId).role(role).build();
```
Why build the `UserModel` every time `getPrincipal` is called?

debanjanc01 commented on skill-tree/src/main/java/com/RDS/skilltree/utils/JWTUtils.java:
```diff
@@ -0,0 +1,88 @@
+    private Claims extractAllClaims(String token) {
+        return Jwts
+                .parser()
+                .setSigningKey(convertToRSAPublicKey(publicRDSKeyString))
```
The value of `convertToRSAPublicKey(publicRDSKeyString)` is always going to be the same as the `publicRDSKeyString` is going to be unchanged.
Rather than using the method call here to convert the key every time `extractAllClaims` is called (which will be for every request that has a Token in header), can we optimise this?

bhtibrewal replied:
was also thinking of it, but didn't get how I can do a calculation and store the result of it here 
 can you please suggest me, something I need to look for to achieve this

debanjanc01 replied:
Class initialization/construction!

debanjanc01 commented on skill-tree/src/main/java/com/RDS/skilltree/Filters/JWTAuthenticationFilter.java:
```diff
@@ -0,0 +1,64 @@
+        /* extract token from header */
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
+            return bearerToken.substring(7);
```
Avoid using magic numbers! I know logically you are reading from the 7th position, but it's better to not have a random number suddenly appearing.

Someone new (with no context) reading the codebase would think how did you come up with 7?

Try using "Bearer ".length().
Also, maintain the strings "Authorization" and "Bearer " as `private static final` class variables.

Also,

bhtibrewal replied:
done, thanks for this suggestion

debanjanc01 commented on skill-tree/src/main/java/com/RDS/skilltree/Filters/JWTAuthenticationFilter.java:
```diff
@@ -0,0 +1,64 @@
+
+
+            if (StringUtils.hasText(token) && jwtUtils.validateToken(token)) {
+                String rdsUserId = jwtUtils.getRDSUserId(token);
+                String role = jwtUtils.getUserRole(token);
```
Rather than parsing the same `token` string multiple times, you can extract the Claims once, and then use the Claims to extract whatever information you need.

The `validateToken` method itself may be used to return the Claims if the token is valid.

debanjanc01 commented on skill-tree/src/main/java/com/RDS/skilltree/Authentication/UserAuthenticationToken.java:
```diff
@@ -0,0 +1,41 @@
+        super(List.of(new SimpleGrantedAuthority(role)));
+
+        this.rdsUserId = rdsUserId;
+        this.role  = UserRole.fromString(role);
```
What happens in the case there is a new role present in the token, and the enum is not updated in code?

bhtibrewal replied:
than this function return null, that means we didn't get the role, and we want the authorisation to stop them

debanjanc01 replied:
But that wouldn't be the behaviour right? The `this.role` would be set to `null`.

bhtibrewal replied:
added a guest role, so if the above scenarios we assign guest role with read-only access

debanjanc01 commented on skill-tree/src/main/java/com/RDS/skilltree/utils/JWTUtils.java:
```diff
@@ -0,0 +1,88 @@
+            X509EncodedKeySpec keySpec = new X509EncodedKeySpec(publicKeyBytes);
+            return (RSAPublicKey) keyFactory.generatePublic(keySpec);
+        } catch (InvalidKeySpecException exception) {
+            throw new RuntimeException("Invalid RSA key");
```
Is this a RuntimeException by design?

debanjanc01 commented on skill-tree/src/main/java/com/RDS/skilltree/Filters/JWTAuthenticationFilter.java:
```diff
@@ -0,0 +1,64 @@
+
+
+                UserAuthenticationToken authentication = new UserAuthenticationToken(role, rdsUserId);
+                authentication.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));
```
No need to create a new object of `WebAuthenticationDetailsSource` every single time. You can also directly use constructor used inside `buildDetails`.

debanjanc01 replied:
@bhtibrewal at line:42 you are creating a new instance of `WebAuthenticationDetailsSource` every single time the method is called. Is that needed?

debanjanc01 reviewed (changes requested):
Functionality looks good. Leaving a few suggestions.

iamitprakash commented on skill-tree/src/main/java/com/RDS/skilltree/Endorsement/EndorsementController.java:
```diff
@@ -57,4 +57,10 @@ public ResponseEntity<GenericResponse<EndorsementDTO>> postEndorsement(@RequestB
 
     }
+
+    // NOTE: dummy endpoint added to check super-user authorisation
```
didn't get this requirement ..?

iamitprakash commented on skill-tree/src/main/java/com/RDS/skilltree/User/UserRole.java:
```diff
@@ -1,8 +1,30 @@
+     UserRole(String label) {
+        this.label = label;
+    }
+    public static UserRole fromString(String text) {
```
The fromString method returns null if no matching role is found. While this is a valid approach, consider adding a custom exception or throwing an IllegalArgumentException if a user tries to convert an invalid string to a UserRole. This can improve code clarity and error handling.

bhtibrewal replied:
added the guest role with read-only as we discussed

iamitprakash commented on skill-tree/src/main/java/com/RDS/skilltree/User/UserRole.java:
```diff
@@ -1,8 +1,30 @@
+        }
+        return null;
+    }
+    public static String[] getAllRoles(){
```
getAllRoles method retrieves all labels using a stream and then converts them back to an array. This can be simplified by directly collecting the labels into an array using mapToArray(String[]::new).

bhtibrewal replied:
tried this but that was not fulfilling my requirement of converting 
```
UserRoles i.e., [{name: "USER", label:"user"}, name: "MEMBER", label:"member"}]
```
to 
```
["USER", "MEMBER",...]
```

iamitprakash commented on skill-tree/src/main/java/com/RDS/skilltree/User/UserRole.java:
```diff
@@ -1,8 +1,30 @@
+import java.util.Arrays;
+import java.util.Collection;
+
+public enum UserRole  {
```
can we add Guest role to handle read only permissions all over application

bhtibrewal commented:
> @bhtibrewal looking at the attached screenshots, I'm curious — should the health endpoint be behind a token-based authentication?

yes you were right /health shouldn't be behind authentication updated it

debanjanc01 commented on skill-tree/src/main/java/com/RDS/skilltree/Filters/JWTAuthenticationFilter.java:
```diff
@@ -0,0 +1,64 @@
+    @Autowired
+    private JWTUtils jwtUtils;
+
+    private static final String bearerString = "Bearer ";
```
```suggestion
    private static final String BEARER_PREFIX = "Bearer ";
```

debanjanc01 replied:
@bhtibrewal naming convention for static final String.

satyam73 commented on skill-tree/src/main/resources/application.properties:
```diff
@@ -9,3 +9,5 @@ API_V1_PREFIX=/api/v1
 management.endpoints.web.exposure.include=health,info,metrics
 logging.level.root=ERROR
+spring.profiles.active=${SPRING_PROFILES_ACTIVE}
+cookieName=rds-session-v2-development
```
We have to make application property file for prod too

satyam73 reviewed (approved):
LGTM as per my knowledge of Java

bhtibrewal commented:
Merging this as this has 2 approvals and I have addressed all the comments I received on the PR, If anything is left I will create another PR to fix it
